[HDR] Tonemap the HDR image when "dynamic-range-limit: standard;" is applied

shallawa opened the pull request:
#### 6b3061ae38db5aa054c87842d19c0b3d90d853bb
<pre>
[HDR] Tonemap the HDR image when &quot;dynamic-range-limit: standard;&quot; is applied
<a href="https://bugs.webkit.org/show_bug.cgi?id=292463">https://bugs.webkit.org/show_bug.cgi?id=292463</a>
<a href="https://rdar.apple.com/150555655">rdar://150555655</a>

Reviewed by Simon Fraser.

An HDR image with &quot;dynamic-range-limit: standard;&quot; should be excluded from the
layer detection for HDR content. When drawing this image, it should also be
tone-mapped manually. The tone-mapping can be done by drawing the HDR image to
un-accelerated ImageBuffer which is created with SDR color-space.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-basic-image-dynamic-range-limit.html: Added.
* LayoutTests/fast/images/hdr-basic-image-dynamic-range-limit-expected.html: Added.
* LayoutTests/fast/images/hdr-basic-image-dynamic-range-limit.html: Added.
* LayoutTests/fast/images/hdr-image-layer-sdr-dynamic-range-limit-expected.html: Added.
* LayoutTests/fast/images/hdr-image-layer-sdr-dynamic-range-limit.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-dynamic-range-limit-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-basic-image-dynamic-range-limit-expected.txt: Added.
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::hasHDRContent const):
(WebCore::NativeImage::drawWithToneMapping):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/NativeImageSource.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::appendImageSourceOption):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):

Canonical link: <a href="https://commits.webkit.org/294643@main">https://commits.webkit.org/294643@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10951f98efa14501a5e81bbaacca23bdfe481d4b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102545 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22214 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12528 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107706 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53182 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104584 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22522 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30720 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/34994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17454 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/92562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/58349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17292 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10590 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52539 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95216 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10659 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110081 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101154 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29677 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30041 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88756 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/86598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22033 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/31416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9143 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29605 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/34917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124788 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29416 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34620 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/32739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30975 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5915,6 +5915,10 @@ static bool hasVisibleBoxDecorationsOrBackground(const RenderElement& renderer)
 static bool rendererHasHDRContent(const RenderElement& renderer)
 {
+    auto& style = renderer.style();
+    if (style.dynamicRangeLimit().toPlatformDynamicRangeLimit() == PlatformDynamicRangeLimit::standard())
```
We shouldn't have to convert to a `PlatformDynamicRangeLimit` to ask if it's `standard`.

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/NativeImage.h:
```diff
@@ -55,9 +55,11 @@ class NativeImage final : public RenderingResource {
     WEBCORE_EXPORT Headroom headroom() const;
 
     void draw(GraphicsContext&, const FloatRect& destRect, const FloatRect& srcRect, ImagePaintingOptions);
+    void drawToneMapping(GraphicsContext&, const FloatRect& destinationRect, const FloatRect& sourceRect, ImagePaintingOptions);
```
Why not just add a flag to `ImagePaintingOptions`? If you keep this, call it `drawWithToneMapping()` or `drawToneMapped()`.

shallawa replied:
The name was changed to `drawWithToneMapping()`.  Having a flag in `ImagePaintingOptions` seems a little bit complicated.

smfr commented on Source/WebCore/platform/graphics/NativeImage.cpp:
```diff
@@ -104,11 +104,45 @@ DestinationColorSpace NativeImage::colorSpace() const
+{
+    auto colorSpaceForToneMapping = [](GraphicsContext& context) {
+#if PLATFORM(IOS_FAMILY)
+        // iOS typically renders into extended range sRGB to preserve wide gamut colors, but we want
```
This comment seems to be conflating "extended range" with "extended dynamic range". P3 is an extended range colorspace (values outside of sRGB).

shallawa replied:
I fixed the comment.

smfr commented on Source/WebCore/platform/graphics/NativeImage.cpp:
```diff
@@ -104,11 +104,45 @@ DestinationColorSpace NativeImage::colorSpace() const
+        return DestinationColorSpace::DisplayP3();
+#else
+        // Otherwise, match the colorSpace of the GraphicsContext.
+        return context.colorSpace();
```
If we use the destination context's colorspace on macOS, it would seem that using the intermediate buffer doesn't introduce a behavior change. Is it just impacting behavior because it's Unaccelerated? If so, that deserves a command and a radar link.

smfr replied:
Also, here in cross-platform code doesn't seem to be the right place to work around Apple platform limitations.

shallawa replied:
Although the destination color space on macOS may extended dynamic range colorspace, the pixel format of the  intermediate buffer is BGRA8. This will force the tone-mapping anyway.

Why is this function is cross-platform? Because all platforms can save memory if `dynamic-range-limit: standard;`. Like what we do for Apple ports, they can create normal 4-bytes-per-pixel buffers for the layers. So the tone-mapping will be needed in this case.

shallawa committed:
[HDR] Tonemap the HDR image when "dynamic-range-limit: standard;" is applied
https://bugs.webkit.org/show_bug.cgi?id=292463
rdar://150555655

Reviewed by Simon Fraser.

An HDR image with "dynamic-range-limit: standard;" should be excluded from the
layer detection for HDR content. When drawing this image, it should also be
tone-mapped manually. The tone-mapping can be done by drawing the HDR image to
un-accelerated ImageBuffer which is created with SDR color-space.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-basic-image-dynamic-range-limit.html: Added.
* LayoutTests/fast/images/hdr-basic-image-dynamic-range-limit-expected.html: Added.
* LayoutTests/fast/images/hdr-basic-image-dynamic-range-limit.html: Added.
* LayoutTests/fast/images/hdr-image-layer-sdr-dynamic-range-limit-expected.html: Added.
* LayoutTests/fast/images/hdr-image-layer-sdr-dynamic-range-limit.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-dynamic-range-limit-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-basic-image-dynamic-range-limit-expected.txt: Added.
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::hasHDRContent const):
(WebCore::NativeImage::drawWithToneMapping):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/NativeImageSource.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::appendImageSourceOption):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):

Canonical link: https://commits.webkit.org/294643@main
